### PR TITLE
Create CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by MistakeNot4892**

### CODE REVIEW

**Comment by ElArGee** [comment]
But I like relative pathing for procs. And the underscore key is so far away...

**Comment by MistakeNot4892** [comment]
Relative pathing more like REElative pathing

I do not overly mind what we go with re: naming, but redoing the existing code to be Caps will be fiddly

**Comment by MistakeNot4892** [comment]
@Loganbacca how about this?

**Comment by PsiOmegaDelta** [comment]
Relative pathing (for anything but vars) is a pain whenever you want to grab for anything tho, in particular procs and their overrides.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+- Proc names should be written in CamelCase.
+
+- Constants and preprocessor macros/defines should be written in UPPERCASE.
+
+- Variable names should be written in snake_case or lowercase.
+
+- Absolute pathing should be used for procs/verbs, and relative pathing should be used for variables. ex.
```
Comment (Techhead0) [inline CONTRIBUTING.md]: Procs, verbs, and datums.

Nothing like...
```
/obj
	foo
		name = "foo"

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+- Proc names should be written in CamelCase.
```
Comment (ElArGee) [inline CONTRIBUTING.md]: Also know as "PascalCase" and "Upper CamelCase"

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+- Constants and preprocessor macros/defines should be written in UPPERCASE.
```
Comment (ElArGee) [inline CONTRIBUTING.md]: Just confirming we are talking about define constants here and not constant variables? i.e. var/const/foo

**Comment by ElArGee** [comment]
Just to clarify, although I like the clean look of relative paths, I'm not serious about using it in this project. It would be a nightmare to implement it now anyway.

CamelCase on the other hand I would like to see just to differentiate vars from procs.
I'm currently in the process of converting the codebase to CamelCase.

File: CONTRIBUTING.md (hunk 1 above)
Comment (MistakeNot4892) [reply CONTRIBUTING.md]: I'll add an explicit reference to atom paths too.

File: CONTRIBUTING.md (hunk 3 above)
Comment (MistakeNot4892) [reply CONTRIBUTING.md]: I actually meant both, seems like it would be more appropriate than mixing constant vars and constant defines.

**Comment by AntimoniumGame** [comment]
Testing webhook

**Comment by PsiOmegaDelta** [comment]
The BYOND ref/codebase appears to use the following setup, if one wants consistency with that:
* snake_case for vars.
* CamelCase for instance/type procs, i.e. ```/icon/proc/SetIntensity()```, ```/atom/proc/MouseDrop()```
* A general mix of snake_case and lowercase for global procs, i.e. ```/proc/get_dist()```, ```/proc/text2num()```, ```/proc/json_encode()```, ```/proc/copytext()```.

**Comment by ElArGee** [comment]
^ I'm happy with the above.

Another thing to add: no same type overrides. 
i.e.
```
/obj/New()
   do something
   ..()

/obj/New()
   do something else
   ..()
```

Child overrides are fine, but same type overrides just get messy and unreliable. So in theory there should only be one proc definition for /obj/New() that would reside in _object.dm. If another system needs to add something to /obj/New() then it should be done in _object.dm, not in a same type override somewhere else in the codebase.

**Comment by MistakeNot4892** [comment]
I'm not sure I completely agree, just because if we end up modularizing the codebase it might be useful to have a stubs.dm file somewhere with the `/bar/proc/foo()` root definitions and then additional logic defined in the various modules.

**Comment by MistakeNot4892** [comment]
New version, pls read and criticise

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+/obj/proc/Foo()
+   return "bar"
+
+/obj/Foo()
+   . = "foo" + ..()   
+```
+
+- Unless an object is going to be directly added to the map, or the object implements specific behavior not covered by other objects, object variation should not be hardcoded. eg. there are not specific object subtypes for heads, groins and chests, but there are 'stance' and 'grasp' subtypes, which are the basis for arms and legs at runtime.
+
+- Under no circumstances should the internal object variables `step_x` or `step_y` be set, either in the code or on the map. Usually if these are set accidentally it is by using the 'nudge' function while mapping. This is to avoid 'chess mode', in which BYOND assumes we have implemented our own gliding system and does not perform any movement interpolation of its own.
```
Comment (ElArGee) [inline CONTRIBUTING.md]: Technically setting `step_x` or `step_y` actually kicks the engine into pixel movement mode, as opposed to assuming we have our own gliding system. Depending on what you set the `step` variables to will determine how big the chess mode increments are. If they aren't set I'm assuming it defaults to the world tile size, thus making every increment 32 pixels, or one turf.

**Comment by ElArGee** [comment]
I'm pretty happy with this 👍 :shipit:

**Pull request merged**